Add PEP 519 support

aragilar opened the pull request:
Closes #5968 and replaces #6772. This should allow pathlib.Path or other PEP 519 objects to be used with matplotlib.

aragilar commented:
I'm not sure why this is failing on windows, suggestions?

jenshnielsen commented:
Looks like it's a permission error where it tries to write the files `IOError: [Errno 13] Permission denied: u'c:\\users\\appveyor\\appdata\\local\\temp\\1\\tmpo9j2ke.svg'`

@janschulz Do you know anything about the permission model on appveyor

janschulz commented:
Appveyor builds run as admin, so I don't think this is the problem. In my experience, such problem arise (on windows) from doing creation/deletions too fast: one process still has the file opened/locked and the next call is not allowed to do something with it.

janschulz commented on lib/matplotlib/tests/test_animation.py:
```diff
@@ -149,6 +156,94 @@ def animate(i):
+
+    # Use NamedTemporaryFile: will be automatically deleted
+    F = tempfile.NamedTemporaryFile(suffix='.' + extension)
+    F.close()
```
This works on windows to get a named temp file name...

janschulz commented on lib/matplotlib/tests/test_backend_pdf.py:
```diff
@@ -132,3 +133,37 @@ def test_grayscale_alpha():
+        with PdfPages(FakeFSPathClass(tmpfile.name)) as pdf:
+            fig, ax = plt.subplots()
+            ax.plot([1, 2], [3, 4])
+            pdf.savefig(fig)
```
... but this doesn't: the NamedTemporaryFile still has the file open and savefig writes to it. The other test closes the file and reuses only the name.

janschulz replied:
in pypandoc I use this:

```
@contextlib.contextmanager
def closed_tempfile(suffix, text=None):
    with tempfile.NamedTemporaryFile('w+t', suffix=suffix, delete=False) as test_file:
        file_name = test_file.name
        if text:
            test_file.write(text)
            test_file.flush()
    yield file_name
    shutil.rmtree(file_name, ignore_errors=True)

[...]

with closed_tempfile('.md', text='#some title\n') as file_name:
    # do something...
```

janschulz replied:
in pypandoc I use this:

```
@contextlib.contextmanager
def closed_tempfile(suffix, text=None):
    with tempfile.NamedTemporaryFile('w+t', suffix=suffix, delete=False) as test_file:
        file_name = test_file.name
        if text:
            test_file.write(text)
            test_file.flush()
    yield file_name
    shutil.rmtree(file_name, ignore_errors=True)

[...]

with closed_tempfile('.md', text='#some title\n') as file_name:
    # do something...
```

aragilar commented:
Updated and rebased on top of current master. I did have to add matplotlib.testing.nose and matplotlib.testing.nose.plugins to the included modules to build locally though, is that expected?

aragilar commented:
Also, is it worth creating a separate pull request with @janschulz's closed_tempfile context manager?

tacaswell commented on lib/matplotlib/tests/test_backend_pdf.py:
```diff
@@ -132,3 +133,39 @@ def test_grayscale_alpha():
+    try:
+        from pathlib import Path
+    except ImportError:
+        raise SkipTest("pathlib not installed")
```
@Kojoley What is the correct way to handle this now?

Kojoley replied:
It is OK to use `SkipTest` as pytest supports it, but `skip` (imported from `from .testing import skip`) is preferable as it does not use nose when you are using pytest.

tacaswell commented:
We are in the process of converting from nose to pytest (and the whole test suite will currently run under both) with almost all of the work being done by @Kojoley . 

Can you do the context manager change in this PR?

aragilar commented:
This should be all good now.

NelleV commented on lib/matplotlib/cbook.py:
```diff
@@ -2701,3 +2705,44 @@ def __exit__(self, exc_type, exc_value, traceback):
+            raise TypeError("expected str, bytes or os.PathLike object, not "
+                            + path_type.__name__)
+
+def fspath_no_except(path):
```
I think this should be a private function.
If it is not a private function, it should be documented.

I also don't really like this name, but I am not coming up with anything better for now…

QuLogic replied:
I seem to recall something (Sphinx? NumPyDoc?) using a `safe_` prefix; so something like `safe_fspath` (or even more explicitly `safe_fspath_to_str`)?

NelleV replied:
That'd be a much better name.

aragilar replied:
I'm not a fan of safe_fspath, as it's not really safe, since the reason it exists to to pass through anything that's not a PEP 519 path, and validating anything else doesn't happen. Also, there's no guarantee that it'll return a string, as __fspath__ can return either bytes or a string (I've special-cased pathlib here as there exist versions of pathlib without PEP 519 support), so safe_fspath_to_str is even worse. How about fspath_passthrough?

NelleV commented on lib/matplotlib/cbook.py:
```diff
@@ -2701,3 +2705,44 @@ def __exit__(self, exc_type, exc_value, traceback):
+                    return six.text_type(path)
+
+            raise TypeError("expected str, bytes or os.PathLike object, not "
+                            + path_type.__name__)
```
That is not pep8 compliant. Please break the line after the +.
I would also prefer using string formatting instead of concatenation.

QuLogic replied:
PEP8 does not prescribe either way, but has a slight preference for the code as it's written.

aragilar replied:
So change or leave?

NelleV commented on lib/matplotlib/tests/test_animation.py:
```diff
@@ -128,20 +134,72 @@ def animate(i):
+
+
+@cleanup
+def check_save_animation_pep_519(writer, extension='mp4'):
```
I would avoid referencing to peps in function name. It is unclear what it does.

aragilar replied:
What about check_save_animation_PathLike?

NelleV replied:
`check_save_animation_path_like`?

NelleV commented on lib/matplotlib/tests/test_animation.py:
```diff
@@ -128,20 +134,72 @@ def animate(i):
+
+
+@cleanup
+def check_save_animation_pathlib(writer, extension='mp4'):
```
This test and the previous one looks like they could be refactoring to loop other Path and fspath compatible object to minimize the code duplication.

NelleV commented on lib/matplotlib/tests/test_backend_pdf.py:
```diff
@@ -132,3 +129,34 @@ def test_grayscale_alpha():
+
+
+@cleanup
+def test_pdfpages_accept_pep_519():
```
Those two tests look like they could be refactored into one.

aragilar replied:
The test_pdfpages_accept_pep_519 and test_savefig_accept_pathlib? I'm not sure it's possible to skip part of a test and have that appear in nose/pytest.

NelleV replied:
mmh… good question. 
@Kojoley  any idea on that?

NelleV commented on lib/matplotlib/tests/test_testing.py:
```diff
@@ -0,0 +1,21 @@
+    text = "This is a test"
+    with closed_tempfile(".txt", text=text) as f:
+        with open(f, "rt") as g:
+            assert text == g.read()
```
Can you add the approriate code to be able to run the tests in this file independantly from the rest?
```python
if __name__=='__main__':
     nose.runmodule(argv=['-s','--with-doctest'], exit=False)
```

aragilar replied:
I'll add this

NelleV commented:
This patch will be useful for the future! Thanks for tackling this work.

aragilar commented on lib/matplotlib/cbook.py:
```diff
@@ -2701,3 +2705,44 @@ def __exit__(self, exc_type, exc_value, traceback):
                     pass
+
+
+try:
```
Sure, I can do that.

aragilar replied:
So I'm not sure what can actually be moved from matplotlib.cbook to matplotlib.cbook.fixes, as the fspath stuff is the only compatibility things I can see, and there's already matplotlib.compat...

QuLogic commented on lib/matplotlib/testing/__init__.py:
```diff
@@ -164,3 +166,21 @@ def setup():
+    """
+    with tempfile.NamedTemporaryFile(
+        'w+t', suffix=suffix, delete=False
+    ) as test_file:
```
```python
    with tempfile.NamedTemporaryFile('w+t', suffix=suffix,
                                     delete=False) as test_file:
```
fits within 79 chars, and looks reasonable, I think.

QuLogic commented on lib/matplotlib/tests/test_animation.py:
```diff
@@ -128,20 +134,72 @@ def animate(i):
+
+    if not animation.writers.is_available(writer):
+        skip("writer '%s' not available on this system"
+                               % writer)
```
It doesn't for me; I don't know why this is passing either.

aragilar commented on lib/matplotlib/testing/__init__.py:
```diff
@@ -164,3 +166,21 @@ def setup():
+@contextmanager
+def closed_tempfile(suffix='', text=None):
+    """
+    Context manager which yields the path to a closed temporary file with the
```
Does publicly available mean should be used outside of matplotlib, or is the recommended option for other parts of matplotlib? I don't see much use outside matplotlib, unless you were already using matplotlib's testing utils.

mruffalo commented:
I see that this is tagged with the "2.1 (next point release)" milestone -- and I don't follow matplotlib development very closely, so I'm not sure when to expect **2.0** to be released. Is there any chance of this (or a stripped-down version of this) functionality being included in a 1.5.x release relatively soon? I assume not, but I don't see too much harm in asking.

I started using Python 3.6 for my analysis code a week or so ago; it's been a while since the last beta, rc1 is due tonight (as I understand it), and the final release is scheduled for 2016-12-16. PEP 519 is the main thing motivating me to install 3.6 on my machines, since I use `pathlib` extensively and it's a pain to wrap every `Path` object in `str()` when using the convenience I/O functions in things like Pandas, matplotlib, and NetworkX (among others). I enjoyed dropping all of these extra calls to `str()`, but found that matplotlib was still rejecting `Path` objects instead of simply passing them to `open` as I expected. This isn't a big deal but is kind of an eyesore now that I can pass `Path` objects to things like `pd.read_pickle`.

More generally, is there a good justification for the current behavior of throwing an error from matplotlib if e.g. `savefig` is given an unknown type? That doesn't seem to really add anything to the simpler behavior of "if it's a file-like object, use it, else pass it to `open`, and let any exceptions thrown there propagate to the caller" -- and that simpler behavior would have entirely avoided the need for this pull request.

tacaswell commented:
There will be no more 1.5.x releases, 2.0 has an rc1 out final hopefully in late December / early January . 2.1 should be something like late Q1/early Q2 2017.

mangecoeur commented:
Is this likely to land soon?

tacaswell commented:
@aragilar I have done the rebase and taken over this PR.  Please move discussion to #8481
